### PR TITLE
Make CI fail if there are clang format errors

### PR DESCRIPTION
**Pull request by vhvb1989**
Won't pass until #1330 is merged first

fixes: #1329

### CODE REVIEW

**Comment by Jinming-Hu** [comment]
This raises the bar of contribution. Will it discourage contributors from community?

**Review by RickWinter — APPROVED** [review]

File: eng/pipelines/templates/jobs/archetype-sdk-client.yml
Comment (ahsonkhan) [inline eng/pipelines/templates/jobs/archetype-sdk-client.yml:102]: Do we need the command option `-style=file`?

> When using -style=file, clang-format for each input file will try to find the .clang-format file located in the closest parent directory of the input file. When the standard input is used, the search is started from the current directory.

We have a global .clang-format file.

File: eng/pipelines/templates/jobs/archetype-sdk-client.yml
Comment (ahsonkhan) [inline eng/pipelines/templates/jobs/archetype-sdk-client.yml:104]: Do we want untracked files to cause CI build failures? Consider borrowing aspects from the C SDK repo, such as `--untracked-files=no`:
https://github.com/Azure/azure-sdk-for-c/blob/0985be23010cb0989b591d92a57d6b276e960b92/eng/pipelines/templates/jobs/archetype-sdk-client.yml#L344-L362

File: eng/pipelines/templates/jobs/archetype-sdk-client.yml
Comment (ahsonkhan) [inline eng/pipelines/templates/jobs/archetype-sdk-client.yml:109]:
```suggestion
        echo "List of files not formatted correctly:"
```

**Comment by ahsonkhan** [comment]
We have a similar approach in C and didn't observe any issues. In fact, given the error message in the CI is clear about what to do to fix it, it has helped reduce unrelated formatting changes in PRs and making them more focused which makes them easier to review. Having it in CI ensures we stay green :)

File: eng/pipelines/templates/jobs/archetype-sdk-client.yml
Comment (vhvb1989) [reply eng/pipelines/templates/jobs/archetype-sdk-client.yml:102]: I am not sure if by not adding it, clang-format will try to use the .clang-format file or not

File: eng/pipelines/templates/jobs/archetype-sdk-client.yml
Comment (vhvb1989) [reply eng/pipelines/templates/jobs/archetype-sdk-client.yml:104]: `grep modified` will filter only modified files.  Untracked won't be there

File: eng/pipelines/templates/jobs/archetype-sdk-client.yml
Comment (ahsonkhan) [reply eng/pipelines/templates/jobs/archetype-sdk-client.yml:102]: > I am not sure if by not adding it, clang-format will try to use the .clang-format file or not

You can test that by changing the clang-format file and see if re-formatting the files causes a change. It works in C.

File: eng/pipelines/templates/jobs/archetype-sdk-client.yml
Comment (ahsonkhan) [reply eng/pipelines/templates/jobs/archetype-sdk-client.yml:104]: Using `grep` on the output from git seems brittle (because the output can change and you may not be able to catch all the changes). Use the built-in `git` command to avoid having to parse and process the output strings.

Comment (vhvb1989) [reply eng/pipelines/templates/jobs/archetype-sdk-client.yml:104]: > because the output can change

??  can you elaborate?

Comment (ahsonkhan) [reply eng/pipelines/templates/jobs/archetype-sdk-client.yml:144]: In the future, Git could change the output of `git status` to use other text words instead of `modified`, or add similar for untracked/other types of changes.

File: eng/pipelines/templates/jobs/archetype-sdk-client.yml
Comment (ahsonkhan) [inline eng/pipelines/templates/jobs/archetype-sdk-client.yml:140]: Just to confirm, is 11 the latest? From what I gather it is:
https://releases.llvm.org/download.html

File: eng/pipelines/templates/jobs/archetype-sdk-client.yml
Comment (vhvb1989) [reply eng/pipelines/templates/jobs/archetype-sdk-client.yml:144]: > In the future, Git could change ...

`--untracked-files=no` could also change then.  =P. Let's not worry about that

File: eng/pipelines/templates/jobs/archetype-sdk-client.yml
Comment (vhvb1989) [reply eng/pipelines/templates/jobs/archetype-sdk-client.yml]: 11 is not available for ubuntu 18 by default.  We need to use 10.

File: eng/pipelines/templates/jobs/archetype-sdk-client.yml
Comment (ahsonkhan) [reply eng/pipelines/templates/jobs/archetype-sdk-client.yml:144]: That can't change. It's a built-in git command and removing it would be considered a breaking change. The other approach is just text output as a result of running a `git` command.

File: eng/pipelines/templates/jobs/archetype-sdk-client.yml
Comment (Jinming-Hu) [reply eng/pipelines/templates/jobs/archetype-sdk-client.yml]: Then can we use ubuntu 20 for this pipeline?

Comment (ahsonkhan) [reply eng/pipelines/templates/jobs/archetype-sdk-client.yml]: That would be a bigger change. We don't have a ubuntu 20 pipeline atm.

Comment (Jinming-Hu) [reply eng/pipelines/templates/jobs/archetype-sdk-client.yml]: Can we use macOS? We can get clang-format 11 easily on macOS.

Comment (Jinming-Hu) [reply eng/pipelines/templates/jobs/archetype-sdk-client.yml]: Or how about just downloading binary from this page https://releases.llvm.org/download.html#11.0.0 and extract from the tarball?

File: eng/pipelines/templates/jobs/archetype-sdk-client.yml
Comment (ahsonkhan) [inline eng/pipelines/templates/jobs/archetype-sdk-client.yml:140]: Why redirect to `/dev/null`? What motivated this? Does that hide some detail from the console output? Is it necessary?

File: eng/pipelines/templates/jobs/archetype-sdk-client.yml
Comment (ahsonkhan) [inline eng/pipelines/templates/jobs/archetype-sdk-client.yml:147]: Just to confirm, if I run this command locally right after this PR is merged, will there be any code formatting difference?
`find sdk/ -regex '.*\.\(cpp\|hpp\)' -exec clang-format -i {} \;`

How can I run a similar command on Windows (VS dev prompt for instance), or does it have to bash? I ran it in bash, and that works.

Not sure if adding the `/sdk` directory is necessary.

File: eng/pipelines/templates/jobs/archetype-sdk-client.yml
Comment (Jinming-Hu) [reply eng/pipelines/templates/jobs/archetype-sdk-client.yml:140]: Can we consider downloading binary directly from https://releases.llvm.org/download.html#11.0.0 to sustain the conclusion we got in today's meeting?

File: eng/pipelines/templates/jobs/archetype-sdk-client.yml
Comment (vhvb1989) [reply eng/pipelines/templates/jobs/archetype-sdk-client.yml]: I just checked and it is the default option. So, it is actually the same. 
I think it is good to have so it is clear

File: eng/pipelines/templates/jobs/archetype-sdk-client.yml
Comment (vhvb1989) [reply eng/pipelines/templates/jobs/archetype-sdk-client.yml:144]: We don't really need to worry about that.
Tell me when was the last time the output changed?

File: eng/pipelines/templates/jobs/archetype-sdk-client.yml
Comment (vhvb1989) [reply eng/pipelines/templates/jobs/archetype-sdk-client.yml:140]: > Why redirect to /dev/null? 

I don't want to show the output of those commands as part of the log. It is irrelevant

Comment (vhvb1989) [reply eng/pipelines/templates/jobs/archetype-sdk-client.yml:140]: > Can we consider downloading binary directly 

Will try it. Makes me a little worry to think if the link could ever go down and break the build. But just a little :P

File: eng/pipelines/templates/jobs/archetype-sdk-client.yml
Comment (vhvb1989) [reply eng/pipelines/templates/jobs/archetype-sdk-client.yml:147]: I think the powerShell supports `find` command.  

However, if you are on VS, you should rely on VS using the .clang-format file to do formatting. 

The /sdk is to speed up a little bit in the case you have already generated the project.  For instance, if you generate debug and release projects on Windows, you will get the folder out/ with all the source files to be built.  Then you will be checking its format.
That's why we only care about ./sdk content

File: eng/pipelines/templates/jobs/archetype-sdk-client.yml
Comment (vhvb1989) [reply eng/pipelines/templates/jobs/archetype-sdk-client.yml]: Turns out there are binaries for Ubuntu 12 and 20 but not for 18  ...  We would need to get source and build which could take some extra time.

Instead, I check with macOS as suggested by @Jinming-Hu  and we get version 11 by default there. So I moved the CI check to be macOS only

File: eng/pipelines/templates/jobs/archetype-sdk-client.yml
Comment (ahsonkhan) [reply eng/pipelines/templates/jobs/archetype-sdk-client.yml:148]: > The /sdk is to speed up a little bit in the case you have already generated the project.

Nice, great idea!

File: eng/pipelines/templates/jobs/archetype-sdk-client.yml
Comment (ahsonkhan) [reply eng/pipelines/templates/jobs/archetype-sdk-client.yml]: > I don't want to show the output of those commands as part of the log. It is irrelevant

Won't it be useful to know the reason if the CI step was to fail due to `sudo apt-get install clang-format-10` (for example if it was because of networking)? Is it a lot of logs that get printed? I don't see any harm in keeping it.

File: eng/pipelines/templates/jobs/archetype-sdk-client.yml
Comment (ahsonkhan) [reply eng/pipelines/templates/jobs/archetype-sdk-client.yml:145]: `git status -s` and `git status --porcelain` are more efficient and are better suited for our purpose here for text processing with scripts, since git status is meant for human readability and git doesn't guarantee that won't change. For example. those options to git status don't output "modified" but rather "M".

https://www.stefanjudis.com/today-i-learned/the-short-version-of-git-status-and-the-close-but-different-porcelain-mode/
> ... the --porcelain flag is used to generate output that is easily parseable by a script. Imagine you run some automation that checks the status of a git repository before running further commands. In that case, you don't want to have help text and terminal colors included in the output.

> Additionally, **you also don't want that the output changes or differs from git version to git version. --porcelain guarantees that there won't be backward-incompatible changes to the output so that your scripts won't break with a git update.**

The same is mentioned in the documentation for git:
https://git-scm.com/docs/git-status#Documentation/git-status.txt---porcelainltversiongt
> Give the output in an easy-to-parse format for scripts. This is similar to the short output, **but will remain stable across Git versions** and regardless of user configuration.

>Tell me when was the last time the output changed?

From what I can infer, it was last changed ~7 years ago, but its probably still better to use the documented and recommended approach, since it _can_ change (i.e. git is allowed to, per their docs):
https://github.com/git/git/blame/master/wt-status.c

That said, I won't push for it any further, and leave it up to you to keep using `grep` and look for `modified` in the output of `git status`, if you feel strongly about using this approach instead.

File: eng/pipelines/templates/jobs/archetype-sdk-client.yml
Comment (ahsonkhan) [reply eng/pipelines/templates/jobs/archetype-sdk-client.yml:143]: > I just checked and it is the default option. So, it is actually the same.

Can you share a source which mentions that's the default value of `style` when that option isn't specified?

File: eng/pipelines/templates/jobs/archetype-sdk-client.yml
Comment (Jinming-Hu) [reply eng/pipelines/templates/jobs/archetype-sdk-client.yml]: Yeah, moving to macOS is a good idea. Also, downloading binary is also an option.

File: eng/pipelines/templates/jobs/archetype-sdk-client.yml
Comment (ahsonkhan) [reply eng/pipelines/templates/jobs/archetype-sdk-client.yml:148]: > However, if you are on VS, you should rely on VS using the .clang-format file to do formatting.

I don't know how to use VS to format all the files in one shot. I tend to run Ctrl+K, Ctrl+D on the files I modify to do the formatting in VS. Got any suggestions on how I can format all files?

**Review by ahsonkhan — APPROVED** [review]

File: eng/pipelines/templates/jobs/archetype-sdk-client.yml
Comment (vhvb1989) [reply eng/pipelines/templates/jobs/archetype-sdk-client.yml:148]: > Got any suggestions on how I can format all files?

Well, when working with VS, why would you need to run formatting to all files?
I think you would need to rely on some 3rd party ext for it, at any case. LIke https://marketplace.visualstudio.com/items?itemName=munyabe.FormatAllFiles

**Comment by vhvb1989** [comment]
@Jinming-Hu  @katmsft  waiting for your approval for this PR

**Review by Jinming-Hu — APPROVED** [review]

File: eng/pipelines/templates/jobs/archetype-sdk-client.yml
Comment (ahsonkhan) [inline eng/pipelines/templates/jobs/archetype-sdk-client.yml:147]: Update the contributing guide doc to say we need clang-format 10 or higher.

https://github.com/Azure/azure-sdk-for-cpp/blob/master/CONTRIBUTING.md#third-party-dependencies
image

**Pull request merged**
